lib/DxcSupport/dxcmem.cpp

```diff
@@ -39,97 +37,56 @@ _Ret_maybenull_ _Post_writable_byte_size_(nBytes) void *DxcThreadAlloc(size_t nB
 void DxcThreadFree(void *) throw();
 
 HRESULT DxcInitThreadMalloc() throw() {
+  DXASSERT(g_ThreadMallocTls == nullptr, "else InitThreadMalloc already called");
   DXASSERT(g_pDefaultMalloc == nullptr, "else InitThreadMalloc already called");
 
-  #ifdef _WIN32
-  DXASSERT(g_ThreadMallocTlsIndex == 0, "else InitThreadMalloc already called");
   // We capture the default malloc early to avoid potential failures later on.
   HRESULT hrMalloc = CoGetMalloc(1, &g_pDefaultMalloc);
   if (FAILED(hrMalloc)) return hrMalloc;
 
-  g_ThreadMallocTlsIndex = TlsAlloc();
-  if (g_ThreadMallocTlsIndex == TLS_OUT_OF_INDEXES) {
-    g_ThreadMallocTlsIndex = 0;
-    g_pDefaultMalloc->Release();
-    g_pDefaultMalloc = nullptr;
-    return E_OUTOFMEMORY;
-  }
-  #else
-  g_pDefaultMalloc = new IMalloc;
-  #endif
+  g_ThreadMallocTls = (llvm::sys::ThreadLocal<IMalloc>*)g_pDefaultMalloc->Alloc(sizeof(llvm::sys::ThreadLocal<IMalloc>));
```

Comment:
Maybe we should handle the out-of-memory case / exception like the existing code does...

IIUC, placement new (line 48) does not actually allocate any memory, so it will not throw an exception. So the out-of-memory case would manifest itself when the return value of Alloc is nullptr...

So, something along these lines:
```cpp
if (g_ThreadMallocTls == nullptr) {
  g_pDefaultMalloc->Release();	
  g_pDefaultMalloc = nullptr;	
  return E_OUTOFMEMORY;
}
```

what do you think?

Reply:
Placement new can't result in null. Internally, ThreadLocal asserts for the extremely rare case of running out of tls keys. This is the only place we use them, so we shouldn't ever need more than one. I encountered the problem because of a bug. It shouldn't happen in practice.

I could add handling of the allocation failing due to running out of memory though.